### PR TITLE
Upgrade vacations/vacation requests.

### PR DESCRIPTION
**Pull request by shayne-fletcher**
Now `v3-v4-upgrade-accept` creates upgrade agreements by accepting v3/v4 upgrade proposals and uses those agreements to upgrade v3 vacation requests and vacations to v4.

### CODE REVIEW

**Comment by hurryabit** [comment]
Is this the code that will collect all existing `Vacation` and `VacationRequest` contracts and push them on the right choices on the `UpgradeAgreement` contracts? If so, it should sit in its own top-level command.

Ultimately, an upgrade will consist of three pieces:
1. A DAML Script that looks for all existing `ExployeeRole` contracts and creates according `UpgradeProposal`s. This script will run with the authority of the company. Currently, this is implemented in `v3-v4-upgrade`. (Maybe `v3-v4-upgrade-init` would be a better name.)
2. A UI component that asks each employee to agree to the `UpgradeProposal` and exercises the `Accept` choice to create an `UpgradeAgreement`. This runs with the employee's authority. Currently, this is what `v3-v4-upgrade-accept` does by pretending to be each employee.
3. A DAML Trigger that waits for newly created `UpgradeAgreement`s and finds all `Vacation` and `VacationRequest` contracts that can be upgraded with this agreement. This trigger will run witht the company's authority only. This should run in a separate command, say `v3-v4-upgrade-finish`.

Does this make sense and does it make clear why I think the new code should live in its own command?

**Comment by shayne-fletcher** [comment]
Sure. No problem.

**Comment by shayne-fletcher** [comment]
I am onboard with these names 👍  `v3-upgrade-init`, `v3-upgrade-accept` and `v3-upgrade-finish`.

**Comment by shayne-fletcher** [comment]
@hurryabit Ready for a last pass review if you are!

File: admin-cli/src/index.ts [hunk 1]
```diff
@@ -70,32 +70,31 @@ namespace v3 {
...
       }
     }
-
-    // Create/approve employee vacation requests.
+    // Create/approve vacation requests.
     for (const employee in config.employees) {
       const employeeInfo = config.employees[employee];
       for (const vacationRequest of employeeInfo.vacationRequests ?? []) {
+        const fromDate = vacationRequest.fromDate;
+        const toDate = vacationRequest.toDate;
```
Comment (hurryabit) [inline admin-cli/src/index.ts]:
```suggestion
        const {toDate, fromDate} = vacationRequest;
```
and remove the line above.

File: admin-cli/src/index.ts [hunk 2]
```diff
@@ -121,32 +120,76 @@ namespace v3v4 {
...
+    // Create update agreements.
+    for (const employeeRoleContract of employeeRoleContracts) {
+      const employee = employeeRoleContract.key;
       const employeeLedger = connect(ledgerParams, employee);
-      const [upgradeProposal] =
-        await employeeLedger.query(davlUpgradev3v4.UpgradeProposal, { employee: employee });
-      await employeeLedger.exercise(
+      const [upgradeProposalContract] =
+        await employeeLedger.query(davlUpgradev3v4.UpgradeProposal, { employee });
+        await employeeLedger.exercise(
```
Comment (hurryabit) [inline admin-cli/src/index.ts]:
```suggestion
      await employeeLedger.exercise(
```
That indentation is confusing.

File: admin-cli/src/index.ts [hunk 3]
```diff
@@ -121,32 +120,76 @@ namespace v3v4 {
...
+    const companyLedger = connect(ledgerParams, company);
+    const employeeUpgradeAgreementContracts =
+      await companyLedger.query(davlUpgradev3v4.UpgradeAgreement, { company });
+    for (const employeeUpgradeAgreementContract of employeeUpgradeAgreementContracts) {
+      const employee = employeeUpgradeAgreementContract.payload.employee;
+      const employeeRoleContract = await companyLedger.lookupByKey(davl4.EmployeeRole, employee);
+      if (employeeRoleContract) {
+        const employeeRole = employeeRoleContract.payload;
+        const vacationRequestContracts = await companyLedger.query(davl3.VacationRequest, { vacation: { employeeRole } });
+        if (vacationRequestContracts.length) {
```
Comment (hurryabit) [inline admin-cli/src/index.ts]:
```suggestion
        if (vacationRequestContracts.length > 0) {
```
Treating numbers as booleans? Too much C, heh? 😃

File: admin-cli/src/index.ts [hunk 4]
```diff
@@ -121,32 +120,76 @@ namespace v3v4 {
...
+          for (const vacationRequestContract of vacationRequestContracts) {
+            const fromDate = vacationRequestContract.payload.vacation.fromDate;
+            const toDate = vacationRequestContract.payload.vacation.toDate;
```
Comment (hurryabit) [inline admin-cli/src/index.ts]:
```suggestion
            const {toDate, fromDate} = vacationRequestContract.payload.vacation;
```
and remove the lone above.

File: admin-cli/src/index.ts [hunk 5]
```diff
@@ -121,32 +120,76 @@ namespace v3v4 {
...
+              { requestId: vacationRequestContract.contractId });
+            console.log(`Upgraded VacationRequest [${fromDate}, ${toDate}] for ${employee}.`);
+          }
+        }
+        const boss = employeeRole.boss;
+        const [bossUpgradeAgreementContract] =
+          await companyLedger.query(davlUpgradev3v4.UpgradeAgreement, { employee: boss, company });
+        if (bossUpgradeAgreementContract) {
+          const vacationContracts = await companyLedger.query(davl3.Vacation, { employeeRole });
+          if (vacationContracts.length) {
```
Comment (hurryabit) [inline admin-cli/src/index.ts]:
```suggestion
          if (vacationContracts.length > 0) {
```

File: admin-cli/src/index.ts [hunk 6]
```diff
@@ -121,32 +120,76 @@ namespace v3v4 {
...
+            for (const vacationContract of vacationContracts) {
+              const fromDate = vacationContract.payload.fromDate;
+              const toDate = vacationContract.payload.toDate;
```
Comment (hurryabit) [inline admin-cli/src/index.ts]: Same as above.

File: test-setup.json [hunk 7]
```diff
@@ -8,7 +8,7 @@
     "Martin": {
       "boss": "Shaul",
       "vacationDays": 25,
-      "acceptProposal": false
+      "acceptProposal": true
```
Comment (hurryabit) [inline test-setup.json]: No, please don't. I would need to adapt my manual testing.

Comment (shayne-fletcher) [reply test-setup.json]: As we agreed in our meeting just now, we leave this in the PR where Martin is changed to `true`, and Robin is now `false` to compensate. Else, we can't upgrade vacation requests.

**Pull request merged**